cpp/src/arrow/array/concatenate.cc

```diff
@@ -248,6 +305,38 @@ class ConcatenateImpl {
     return ConcatenateImpl(child_data, pool_).Concatenate(&out_->child_data[0]);
   }
 
+  template <typename T>
+  enable_if_list_view<T, Status> Visit(const T& type) {
+    using offset_type = typename T::offset_type;
+    out_->buffers.resize(3);
+    out_->child_data.resize(1);
+
+    // Calculate the ranges of values that each list-view array uses
+    std::vector<Range> value_ranges;
+    value_ranges.reserve(in_.size());
+    for (const auto& input : in_) {
+      ArraySpan input_span(*input);
+      Range range;
+      ARROW_ASSIGN_OR_RAISE(std::tie(range.offset, range.length),
+                            list_util::internal::RangeOfValuesUsed(input_span));
+      value_ranges.push_back(range);
```

Comment:
(Not actually a suggestion for this PR, just speculation about future helpers) what would you think of adding a macro for `ASSIGN_OR_RAISE` with structured binding?
```suggestion
      ARROW_ASSIGN_OR_RAISE_LIST((offset, length),
                                 list_util::internal::RangeOfValuesUsed(input_span));
      value_ranges.push_back({offset, length});
```

Reply:
I don't mind the explicitness of `std::tie(range.offset, range.length)` as pairs (which I use a lot) have the problem that you need to understand the context to be sure about the meaning of the `.second` field — is it a length or the inclusive offset bound.